File: Module 1 Assignment Highridge Construction Company Payment Slips.py
worker = []
staff = worker

# Introduction of loop to geerate 400 workers
for i in range(1, 401):
    staff_id = f"staff_{i}" # Generating staff id such as Staff_1, Staff_2 etc
    staff.append(staff_id)

# Print list of 400 Staff
print(staff[:401]) #Displayong list of 400 staff dynamicaly generated
print(f"Total number of staff: {len(staff)}") # Verifying the total count of staff generated

# Payslip Dummy data import
import random
from datetime import date

# Number of staff to generate
staff = 400

# Function to generate a staff payment slip
def generate_payment_slip(staff_id, name, salary, employee_level):
    try:
        # Get today's date for the payment slip
        payment_date = date.today()

        # Format for the payment slip
        payment_slip = f"""
        PAYMENT SLIP
        -------------------
        Worker ID: {staff_id}
        Name: {name}
        Salary: ${salary:,.2f}
        Employee Level: {employee_level}
        Payment Date: {payment_date}
        -------------------
        """
        return payment_slip
    except Exception as e:
        print(f"Error generating payment slip for {staff_id}: {e}")
        return None

# Generate staff list dynamically
staffs = []
for i in range(1,401):
    try:
        staff_id = f"staff_{i}"  # Unique staff ID
        name = f"Staff {i}"  # Staff naming convention
        gender = random.choice(["Male", "Female"])  # Random gender selection
        salary = random.randint(7500, 30000)  # Random salary between $6,000 and $50,000

        # Assign employee level based on conditions (salary Scale, Gender and employee group)
        if 10000 < salary < 20000:
            employee_level = "A1"
        elif 7500 < salary < 30000 and gender == "Female":
            employee_level = "A5-F"
        else:
            employee_level = "General"  # Default level for workers not matching the conditions

        # Add employee to staff list
        staff = {"staff_id": staff_id, "name": name, "salary": salary, "employee_level": employee_level, "gender": gender}
        staffs.append(staff)

    except Exception as e:
        print(f"Error generating staff {i}: {e}")

# Generate payment slips for all staff and store them
payment_slips = []
for staff in staffs:
    try:
        slip = generate_payment_slip(staff["staff_id"], staff["name"], staff["salary"], staff["employee_level"])
        if slip:  # Append only if the payment slip is generated successfully
            payment_slips.append(slip)
    except Exception as e:
        print(f"Error generating payment slip for {staff['staff_id']}: {e}")

# Print the 400 staff payment slips
for i in range(400):  # Print 400 staff payment slips for brevity
    try:
        print(payment_slips[i])
    except IndexError:
        print("Error: Payment slip index out of range.")
        

File: Module 1 Assignment Highridge Construction Company Payment Slips.R
# Initialize an empty vector to store the staff
worker <- c()
staff <- worker

# Introduction of loop to generate 400 staff
for (i in 1:400) {
  staff_id <- paste("staff", i, sep = "_")  # Generating staff ID such as staff_1, staff_2, etc.
  staff <- c(staff, staff_id)  # Append staff ID to the vector
}

# Print list of 400 Staff
print(staff)  # Displaying list of 400 staff dynamically generated

# Verifying the total number of staff generated
cat("Total number of staff:", length(staff), "\n")  # Print total count of staff generated

# Load necessary libraries
library(lubridate)

# Number of staff to be generate
staffs <- 400

# Function to generate 400 staff payment slip
generate_payment_slip <- function(staff_id, name, salary, employee_level) {
  tryCatch({
    # Get today's date for payment slip generation
    payment_date <- Sys.Date()
    
    # Format the payment slip
    payment_slip <- paste(
      "PAYMENT SLIP\n",
      "-------------------\n",
      paste("staff ID:", staff_id, "\n"),
      paste("Name:", name, "\n"),
      paste("Salary: $", format(salary, big.mark = ",", scientific = FALSE), "\n", sep = ""),
      paste("Employee Level:", employee_level, "\n"),
      paste("Payment Date:", payment_date, "\n"),
      "-------------------\n",
      sep = ""
    )
    return(payment_slip)
  }, error = function(e) {
    message(paste("Error generating payment slip for", staff_id, ":", e))
    return(NULL)
  })
}

# Initialize the staff list
staffs <- list()

# Generate 400 staff list dynamically
for (i in 1:400) {
  tryCatch({
    staff_id <- paste("staff", i, sep = "_")  # Unique staff ID
    name <- paste("Staff", i)  # Staff naming convention
    gender <- sample(c("Male", "Female"), 1)  # Gender selection
    salary <- sample(7500:30000, 1)  # salary range between $7,500 and $30,000
    
    # Assign Employee Level based on conditions statement
    if (salary > 10000 & salary < 20000) {
      employee_level <- "A1"
    } else if (salary > 7500 & salary < 30000 & gender == "Female") {
      employee_level <- "A5-F"
    } else {
      employee_level <- "General"  # employee level for staff not matching the conditions statement
    }
    
    # Add employee to staff list
    staff <- list(staff_id = staff_id, name = name, salary = salary, 
                  employee_level = employee_level, gender = gender)
    staffs[[i]] <- staff
  }, error = function(e) {
    message(paste("Error generating staff", i, ":", e))
  })
}

# Generate payment slips for all staff and store them
payment_slips <- list()

for (staff in staffs) {
  tryCatch({
    slip <- generate_payment_slip(staff$staff_id, staff$name, staff$salary, staff$employee_level)
    if (!is.null(slip)) {
      payment_slips <- append(payment_slips, list(slip))
    }
  }, error = function(e) {
    message(paste("Error generating payment slip for", staff$staff_id, ":", e))
  })
}

# Print 400 payment slips
for (i in 1:min(400, length(payment_slips))) {
  tryCatch({
    cat(payment_slips[[i]], "\n")
  }, error = function(e) {
    message("Error: Unable to print payment slip.")
  })
}

# Print total staff count
cat("Total number of staff: ", length(staffs), "\n")

